bittide/tests/Tests/Axi4.hs

```diff
@@ -0,0 +1,216 @@
+-- SPDX-FileCopyrightText: 2023 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=0 #-}
+
+{-# HLINT ignore "Functor law" #-}
+{-# HLINT ignore "Used otherwise as a pattern" #-}
+
+module Tests.Axi4 where
+
+import Clash.Prelude
+
+import Clash.Hedgehog.Sized.Unsigned
+import Clash.Hedgehog.Sized.Vector (genVec)
+import Clash.Sized.Vector(unsafeFromList)
+import Data.Maybe
+import Hedgehog
+import Protocols
+import Protocols.Axi4.Stream
+import Protocols.Wishbone
+import Test.Tasty
+import Test.Tasty.Hedgehog
+
+import Bittide.Axi4
+import Bittide.Extra.Maybe
+import Bittide.SharedTypes
+import Tests.Shared
+
+import qualified Data.List as L
+import qualified GHC.TypeNats as TN
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+axi4Group :: TestTree
+axi4Group = testGroup "Axi4Group"
+  [ testPropertyNamed
+    "Axi4Stream upscaling does not affect packet content" "axisFromByteStreamUnchangedPackets"
+    axisFromByteStreamUnchangedPackets
+  , testPropertyNamed
+    "Axi4Stream downscaling does not affect packet content" "axisToByteStreamUnchangedPackets"
+    axisToByteStreamUnchangedPackets
+  ]
+
+type Packet = [Unsigned 8]
+type BasicAxiConfig nBytes = 'Axi4StreamConfig nBytes 0 0
+
+genAxisM2S :: KnownAxi4StreamConfig conf => Gen userType -> Gen (Axi4StreamM2S conf userType)
+genAxisM2S genUser = do
+  _tdata <- genVec $ genUnsigned Range.constantBounded
+  _tkeep <- genVec Gen.bool
+  _tstrb <- genVec Gen.bool
+  _tlast <- Gen.bool
+  _tid   <- genUnsigned Range.constantBounded
+  _tdest <- genUnsigned Range.constantBounded
+  _tuser <- genUser
+
+  pure $ Axi4StreamM2S{..}
+
+-- | Generate a `axisToByteStream` component with variable input bus width and test
+-- if a stream of multiple generated `Packet`s can be routed through it without being changed.
+axisToByteStreamUnchangedPackets :: Property
+axisToByteStreamUnchangedPackets = property $ do
+  busWidth <- forAll @_ @Integer $ Gen.enum 1 8
+  nrOfPackets <- forAll $ Gen.enum 1 4
+  packetLengths <- forAll $ Gen.list (Range.singleton nrOfPackets) $ Gen.enum 1 16
+  packets <- forAll $ traverse
+    (flip Gen.list (genUnsigned @_ @8 Range.constantBounded) . Range.singleton)
+    packetLengths
+  repeatingReadyList <- forAll $ Gen.filter or $ Gen.list (Range.linear 1 8) Gen.bool
+  case
+    TN.someNatVal $ fromIntegral (busWidth - 1) of
+    SomeNat (succSNat . snatProxy -> _ :: SNat busWidth) -> do
+      let
+        axiStream = Nothing : L.concatMap (packetToAxiStream (SNat @busWidth)) packets
+        axiSlave :: Circuit (Axi4Stream System (BasicAxiConfig 1) ()) (Axi4Stream System (BasicAxiConfig 1) ())
```

Comment:
```suggestion
        axiSlave :: 
          Circuit
            (Axi4Stream System (BasicAxiConfig 1) ())
            (Axi4Stream System (BasicAxiConfig 1) ())
```

Bit of vertical space, but it doesn't exceed 80/90 and it is immediately obvious what it is.